plotly/tools.py

```diff
@@ -1269,3 +1276,547 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+    x and y are the coordinates of the arrow locations.
+    u and v give the x and y components of the arrow vectors.
+
+    Example 1:
+    ```
+    # Just 1 Arrow from (0,0) to (1,1)
+    data = Quiver(x=[0], y=[0], u=[1], v=[1], scale=1)
+    ```
+
+    Example 2:
+    ```
+    x,y = np.meshgrid(np.arange(0,np.pi,.2),np.arange(0,np.pi,.2) )
+    u = np.cos(x)
+    v = np.sin(y)
+    data = Quiver(x, y, u, v)
+    ```
+
+    Example 3:
+    ```
+    x,y = np.meshgrid(np.arange(0,2,.2),np.arange(0,2,.2))
+    u = np.cos(x)*y;
+    v = np.sin(x)*y;
+    data = Quiver(x, y, u, v)
+    ```
+
+    Keywords arguments with constant defaults:
+
+    scale (kwarg, float in [0,1], default = .1):
+        Scales size of the arrows (ideally to avoid overlap).
+
+    angle (kwarg, angle in radians, default = np.pi/9):
+        Angle of arrowhead.
+
+    arrow_scale (kwarg, float in [0,1], default = 0.3):
+        Value multiplied to length of barb to get length of arrowhead.
+
+    barb_color (kwarg, color string, default = color):
+        Set color of barbs.
+
+    arrow_color (kwarg, color string, default = color):
+        Set color of arrow.
+
+    barb_width (kwarg, int greater than or equal to 1, default = width):
+        Change width of lines for barbs.
+
+    arrow_width (kwarg, int greater than or equal to 1, default = width):
+        Change width of lines for arrows.
+    """
+
+    if _numpy_imported is False:
+        raise Exception("To use Quiver() please import numpy as np")
+
+    if len(x) != len(y):
+        raise Exception("x, y, u, and v should be ndarrays of equal dimmension"
+                        "or lists where x and y are the same length"
+                        "and u and v are the same length")
+
+    if len(u) != len(v):
+        raise Exception("x, y, u, and v should be ndarrays of equal dimmension"
+                        "or lists where x and y are the same length"
+                        "and u and v are the same length")
+
+    VALID_KWARGS = ['angle', 'scale',
+                    'arrow_scale', 'barb_color',
+                    'arrow_color', 'barb_width',
+                    'arrow_width']
+    for key in kwargs.keys():
+        if key not in VALID_KWARGS:
+            raise Exception("Invalid keyword argument: '{0}'".format(key))
+
+    # Make x y u v into np.array in case user entered list
+    x = np.array(x)
+    y = np.array(y)
+    u = np.array(u)
+    v = np.array(v)
+
+    # Make arrays of x & y start values
+    Xstart = x.flatten('F')
+    Ystart = y.flatten('F')
+
+    # Make arrays of x & y end values
+    Xend = x + (u * scale)
+    Xend = Xend.flatten('F')
+    Yend = y + (v * scale)
+    Yend = Yend.flatten('F')
+
+    # Make array of nans
+    # These are used as spaces so the separate
+    # arrows can be included in 1 trace
+    nnn = np.empty((len(Xstart)))
+    nnn[:] = np.NAN
+
+    # Combine arrays into matrix
+    Xvals = np.matrix([Xstart, Xend, nnn])
```

Comment:
it looks like you rely *heavily* on numpy, which is fine, this can move to *optional* functionality for now, we can always change it in a non-backwards-incompatibility-breaking way later : D